File: implementmagicdictionary.js
//Objective is to design a 'magic dictionary', where you can add words into the dictionary
//and, given that you can change only one character of the word, if it exists in the dictionary


//Design of a Magic Dictionary using a hashmap to keep track of each word's length

class MagicDictionary {
    contructor() {
        this.map = {}
    }

    buildDict(dict) {
        for (let word of dict) {
            if (!this.map[word.length]) {
                this.map[word.length] = []
            }
            this.map[word.length].push(word)
        }
    }

    search(word) {
        if (!this.map[word.length]) {
            return false
        }

        return this.map[word.length].some(find => {
            let count = 0
            for (let i = 0; i < find.length; i++) {
                if (word[i] !== find[i]) {
                    count++
                }
            }

            return count == 1
        })
    }
}

let dict = new MagicDictionary()
dict.buildDict(["hello", "leetcode"])
dict.search("hello")
dict.search("hhllo")
dict.search("hell")
dict.search("leetcoded")

File: implementmagicdictionary.java
//Java Solution

class MagicDictionary {

    /** Initialize your data structure here. */
    Map<Integer, ArrayList<String>> map;
    public MagicDictionary() {
        map = new HashMap<>();
    }
    
    public void buildDict(String[] dictionary) {
        for (String word : dictionary) {
            map.computeIfAbsent(word.length(), x -> new ArrayList()).add(word);
        }
    }
    
    public boolean search(String searchWord) {
        if (!map.containsKey(searchWord.length())) {
            return false;
        }
        
        for (int i = 0; i < map.get(searchWord.length()).size(); i++) {
            String word = map.get(searchWord.length()).get(i);
            int count = 0;
            
            for (int j = 0; j < word.length(); j++) {
                if (word.charAt(j) != searchWord.charAt(j)) {
                    if (++count > 1) {
                        break;
                    }
                }
            }
            
            if (count == 1) {
                return true;
            }
        }
        
        return false;
    }
}